### PR TITLE
Ada belief

### PR DESCRIPTION
**Pull request by quickwritereader**
## What changes were proposed in this pull request?

AdaBelief updater: it was agreed to modify changes on the copy of AdamUpdater. This way we can improve it later.

## How was this patch tested?

unit test

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by quickwritereader** [comment]
I just added one test case in cpp:
here is the python code that was used to get test numbers:
```
import numpy as np
alpha=0.001
beta1=0.9
beta2=0.999
epsilon=1.e-8
#https://arxiv.org/pdf/2010.07468.pdf
def update( t, w, gradW,   mt, st):
        mt =  beta1* mt + (1- beta1)*gradW 
        st =  beta2* st + (1- beta2)*((gradW-mt)**2) + epsilon
        mt_corr =  mt/(1- beta1**t) 
        st_corr =  st/(1- beta2**t)
        upW= alpha*(mt_corr/(np.sqrt(st_corr)+epsilon)) 
        w = w -  upW 
        return ( w, upW,  mt, st )
#if you want to test with more precision np.set_printoptions(precision=9)
grad = np.array([1,2,3,4,5], dtype = np.float32)
w=np.zeros(5, dtype = np.float32)
mt=np.zeros(5, dtype = np.float32)
st = np.zeros(5, dtype = np.float32)
for t in range(1,4):
    w, upW,    mt, st = update(t,w,grad, mt,st )
    print(f"---{t}----")
    print(f"update {upW}")
    print(f" s state {st} ")    
    print(f" m state {mt} ")
```

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/config/AdaBelief.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * AdaBelief 
+ * https://arxiv.org/pdf/2010.07468.pdf
+ */
+@Data
+@Builder(builderClassName = "Builder")
+public class AdaBelief implements IUpdater {
+
+    public static final double DEFAULT_LEARNING_RATE = 1e-3;
+    public static final double DEFAULT_EPSILON = 1e-8;
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/config/AdaBelief.java]: The paper authors have found that a smaller default epsilon works better:

> Epsilon in AdaBelief is different from Adam (typically eps_adabelief = eps_adam*eps_adam)
> ( eps of Adam in Tensorflow is 1e-7, in PyTorch is 1e-8, need to consider this when use AdaBelief in Tensorflow)

So I guess we might want to use either 1e-14 or 1e-16 here.

**Comment by treo** [comment]
Other than the epsilon thing, it looks good to me. 

I've only read through the code though and didn't run it. But the test looks reasonable.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/config/AdaBelief.java (hunk 1 above)
Comment (quickwritereader) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/config/AdaBelief.java]: changed to 1e-14

**Comment by quickwritereader** [comment]
> Other than the epsilon thing, it looks good to me.
> 
> I've only read through the code though and didn't run it. But the test looks reasonable.

let's not rush then. it needs to be checked again properly.

**Review by agibsonccc — APPROVED** [review]

**Pull request merged**